File: sdks/java/src/main/java/com/canarygate/SseParser.java
package com.canarygate;

import java.util.ArrayList;
import java.util.List;

public final class SseParser {

    private SseParser() {
    }

    public record SseEvent(String event, String data, Integer retryMs) {
    }

    public static SseEvent parseBlock(String block) {
        String event = "message";
        List<String> dataLines = new ArrayList<>();
        Integer retryMs = null;

        for (String line : block.split("\\r?\\n")) {
            if (line.isEmpty() || line.startsWith(":")) continue;

            int separatorIndex = line.indexOf(':');
            String field = separatorIndex == -1 ? line : line.substring(0, separatorIndex);
            String value = separatorIndex == -1 ? "" : line.substring(separatorIndex + 1).stripLeading();

            switch (field) {
                case "event" -> event = value.isEmpty() ? "message" : value;
                case "data" -> dataLines.add(value);
                case "retry" -> {
                    try {
                        int parsed = Integer.parseInt(value);
                        if (parsed > 0) {
                            retryMs = parsed;
                        }
                    } catch (NumberFormatException ignored) {
                    }
                }
                default -> {
                }
            }
        }

        if (dataLines.isEmpty() && retryMs == null) {
            return null;
        }
        return new SseEvent(event, String.join("\n", dataLines), retryMs);
    }
}

File: sdks/csharp/src/CanaryGate/SseParser.cs
using System;
using System.Collections.Generic;
using System.Globalization;

namespace CanaryGate;

public sealed record SseEvent(string Event, string Data, int? RetryMs);

public static class SseParser
{
    public static SseEvent? ParseBlock(string block)
    {
        var eventName = "message";
        var dataLines = new List<string>();
        int? retryMs = null;

        foreach (var line in SplitLines(block))
        {
            if (line.Length == 0 || line.StartsWith(':'))
            {
                continue;
            }

            var separatorIndex = line.IndexOf(':');
            var field = separatorIndex == -1 ? line : line[..separatorIndex];
            var value = separatorIndex == -1 ? string.Empty : line[(separatorIndex + 1)..].TrimStart();

            if (field == "event")
            {
                eventName = value.Length > 0 ? value : "message";
                continue;
            }

            if (field == "data")
            {
                dataLines.Add(value);
                continue;
            }

            if (field == "retry" &&
                int.TryParse(value, NumberStyles.Integer, CultureInfo.InvariantCulture, out var parsedRetry) &&
                parsedRetry > 0)
            {
                retryMs = parsedRetry;
            }
        }

        if (dataLines.Count == 0 && !retryMs.HasValue)
        {
            return null;
        }

        return new SseEvent(eventName, string.Join("\n", dataLines), retryMs);
    }

    private static string[] SplitLines(string block)
    {
        return block.Split(new[] { "\r\n", "\n", "\r" }, StringSplitOptions.None);
    }
}
